mava/systems/mat/anakin/mat.py

```diff
@@ -0,0 +1,597 @@
+                _update_minibatch, (params, opt_state, entropy_key), minibatches
+            )
+
+            update_state = params, opt_state, traj_batch, advantages, targets, key
+            return update_state, loss_info
+
+        update_state = params, opt_state, traj_batch, advantages, targets, key
+
+        # UPDATE EPOCHS
+        update_state, loss_info = jax.lax.scan(
+            _update_epoch, update_state, None, config.system.ppo_epochs
+        )
+
+        params, opt_state, traj_batch, advantages, targets, key = update_state
+        learner_state = LearnerState(params, opt_state, key, env_state, last_timestep)
+
+        metric = traj_batch.info
+
+        return learner_state, (metric, loss_info)
+
+    def learner_fn(learner_state: LearnerState) -> ExperimentOutput[LearnerState]:
+        """Learner function.
+
+        This function represents the learner, it updates the network parameters
+        by iteratively applying the `_update_step` function for a fixed number of
+        updates. The `_update_step` function is vectorized over a batch of inputs.
+
+        Args:
+            learner_state (NamedTuple):
+                - params: The initial model parameters.
+                - opt_state: The initial optimiser state.
+                - key: The random number generator state.
+                - env_state: The environment state.
+                - timesteps: The initial timestep in the initial trajectory.
+        """
+
+        batched_update_step = jax.vmap(_update_step, in_axes=(0, None), axis_name="batch")
+
+        learner_state, (episode_info, loss_info) = jax.lax.scan(
+            batched_update_step, learner_state, None, config.system.num_updates_per_eval
+        )
+        return ExperimentOutput(
+            learner_state=learner_state,
+            episode_metrics=episode_info,
+            train_metrics=loss_info,
+        )
+
+    return learner_fn
+
+
+def learner_setup(
+    env: MarlEnv, keys: chex.Array, config: DictConfig
+) -> Tuple[LearnerFn[LearnerState], Any, LearnerState]:
+    """Initialise learner_fn, network, optimiser, environment and states."""
+    # Get available TPU cores.
+    n_devices = len(jax.devices())
+
+    # Get number of agents.
+    config.system.num_agents = env.num_agents
+
+    # PRNG keys.
+    key, actor_net_key = keys
+
+    # Initialise observation: Obs for all agents.
+    init_x = env.observation_spec().generate_value()
+    init_x = tree.map(lambda x: x[None, ...], init_x)
+
+    _, action_space_type = get_action_head(env)
+
+    if action_space_type == "discrete":
+        init_action = jnp.zeros((1, config.system.num_agents), dtype=jnp.int32)
+    elif action_space_type == "continuous":
+        init_action = jnp.zeros((1, config.system.num_agents, env.action_dim), dtype=jnp.float32)
+    else:
+        raise ValueError("Invalid action space type")
+
+    # Define network and optimiser.
+    actor_network = MultiAgentTransformer(
```

Comment:
can we rename this to mat_network (very optional)

Reply:
I think `MultiAgentTransformer` is a bit nicer. So will keep as is.